### ISSUE TITLE
Entry cannot be found by its draft ID after editing during live preview

### ISSUE THREAD

**Issue by bencresty** [craftcms/cms#9761]
### Description

Previously we needed to request current data during the preview (where a user edits the data) by a different entry ID than the actual live entry, which makes sense as it's a preview draft after editing. This was working before. But this isn't working anymore while not having changed the code. Craft has been upgraded to 3.7.x though, so this seem to have something to do with one of the 3.7 upgrades.

Anything changed on queries and/or preview draft IDs we should know about?

Using the element-api.php this is the main rule telling no entry has been found with the params. If we leave 'id' out of the query it does return something, so for some reason the preview cannot be found by its draftID anymore.

``` php
// ...
'api/<locale>/prevmapmarker/<id>' => function($locale, $id) {
            return [
                'elementType' => Entry::class,
                'paginate' => false,
                'criteria' => [
                    'site' => Site::find()->where(['language' => $locale])->asArray()->one(),
                    'section' => 'mapMarker',
                    'id' => $id,
                    'status' => null,
                ],
                'one' => true,
                'transformer' => function(Entry $entry) {
                    $mapMarkersGroup = $entry->mapMarkersGroup->one();
                    $filteredMapMarkersGroupKey = $mapMarkersGroup !== null
                        ? $mapMarkersGroup->mapMarkersGroupKey
                        : null;

                    return [
                        'id' => $entry->id,
                        'groupKey' => $filteredMapMarkersGroupKey,
                        'spriteKey' => $entry->mapMarkerSpriteKey->value,
                        'loc' => [
                            'x' => count($entry->location) > 0
                                ? intval($entry->location[0]['xLoc'], 10)
                                : null,
                            'y' => count($entry->location) > 0
                                ? intval($entry->location[0]['yLoc'], 10)
                                : null,
                        ],
                    ];
                },
                'meta' => [
                    'requestedId' => intval($id),
                ],
                'serializer' => 'jsonfeed',
                'pretty' => true,
                'cache' => false,
            ];
        },
```

### Additional info

- Craft version: 3.7.7
- PHP version: 3.4.7
- Database driver & version: mysql
- Plugins & versions: Element API 2.7.0

**Comment by brandonkelly** [comment]
Two things to check:

- Make sure that you are passing the `token` query string param along to your Element API request, either via a `token` query string param or an `X-Craft-Token` header.
- Make sure you are passing the entry’s _canonical ID_ to that `<id>` segment in the Element API endpoint’s URL pattern, accessed by either `entry.sourceId` (deprecated) or `entry.canonicalId`.

For example:

```js
(async () => {
  const entryId = {{ entry.canonicalId }};
  const params = Object.fromEntries((new URLSearchParams(window.location.search)).entries());

  const response = await fetch(`/api/entry/${entryId}.json`, {
    headers: {
      'X-Craft-Token': params.token,
    },
  });
  console.log(await response.json());
})();
```

If you follow those, the token will tell Craft to swap out the canonical entry with the draft that the token is associated with.

(The only thing about this that has changed in 3.7 is that `sourceId` is deprecated and `canonicalId` was added to replace it – but they are just aliases of each other. Looking at your code, there’s nothing in there explicitly draft-related, so I’m not really sure how auto-created drafts in < 3.7 would have worked, unless you were already doing something similar to the above.)

In testing this I did uncover a bug in Element API, which is that cached endpoints were responding with the cache even if a token was present on the request. I’ve just fixed that for the next release (https://github.com/craftcms/element-api/commit/5d9dc77934ff67d31c1d6bc2b4666ead7d4eb076).

**Comment by bencresty** [comment]
Hi @brandonkelly , thanks for your response and actions. 

I don't like to see you prematurely closed this issue though, while the issue is still there and not solved. I just found the real problem source now and the conclusion is something indeed did change in Craft to cause this issue.

It would have been nice if you could be a little more careful in assuming things, read my posts a little better first, and verify things being solved before closing with a user mentioning the issue on github here with a good reason. Thanks.

Please see below.

> Two things to check:
> 
> * Make sure that you are passing the `token` query string param along to your Element API request, either via a `token` query string param or an `X-Craft-Token` header.
> * Make sure you are passing the entry’s _canonical ID_ to that `<id>` segment in the Element API endpoint’s URL pattern, accessed by either `entry.sourceId` (deprecated) or `entry.canonicalId`.
> 
> For example:
> ...
> If you follow those, the token will tell Craft to swap out the canonical entry with the draft that the token is associated with.

As written in my posts above this was all working before for months and I didn't change any code, only upgraded craft and plugins. Meaning clearly I used a token, otherwise it would never have worked to get data from the changes done while the live preview was open.

Unfortunately you didn't read my posts above carefully because if you did you would have known I don't use either `sourceId` nor `canonicalId`, but the `{id}` tag inside the Preview Targets field in the CP which I pass from front-end to the element-api as `id` for given 'action'. I do that like this in the PRO version of the CP:

![preview-targets](https://user-images.githubusercontent.com/69322733/131461519-898314d5-81a2-4982-9d1c-d69c14c27a65.png)

This still doesn't work now, also after just upgrading to Craft 3.7.10. The new version of element-api isn't up yet on composer, so I didn't upgrade that yet, but I don't have to as I found the real issue and it's not element-api:

Because when I change `{id}` with `{canonicalId}` in the Preview Targets field in the CP it is working as expected and as before again.

**Meaning the real issue here is that the value of the `{id}` tag in Craft isn't the same as before and now represents another value, which is clearly not the ID of the preview draft anymore when the live preview is open and something has been edited by the user, but some other value which doesn't make sense to me as it doesn't seem to be the live entry ID either.

When you let the `{id}` tag point to the `{canonicalid}` internally in Craft, at least when the preview is open and an edit has been made by the user, it should work as before again and so solve this issue.
**

> (The only thing about this that has changed in 3.7 is that `sourceId` is deprecated and `canonicalId` was added to replace it – but they are just aliases of each other. Looking at your code, there’s nothing in there explicitly draft-related, so I’m not really sure how auto-created drafts in < 3.7 would have worked, unless you were already doing something similar to the above.)

Again, I don't use the `sourceId` nor the `canonicalId`, but the `{id}` tag in the CP, which was working before. So it's clear that the tag now got a different value, which is a mistake. The `{id}`-tag should still return the same ID-value as before, which should still be the ID of the draft when the preview is open and something is edited in the entry by the user.

> In testing this I did uncover a bug in Element API, which is that cached endpoints were responding with the cache even if a token was present on the request. I’ve just fixed that for the next release ([craftcms/element-api@5d9dc77](https://github.com/craftcms/element-api/commit/5d9dc77934ff67d31c1d6bc2b4666ead7d4eb076)).

Always nice to fix other bugs, so cool. That clearly wasn't solving this issue though.

**Comment by bencresty** [comment]
Forget my last post. It's not working with the `canonicalId` either.

After lots of testing again it's clear again here that something has changed is in the element api (meaning the build in craft api, not the element-api plugin).

**This is what happens:**

1) Live entry ID is 84589. When using the `{id}` var in the preview targets as url param, it returns 84589, because the entry isn't edited yet.

2) When requesting a live preview the live entry ID (84589) gets sent to the frontend, which in turn does nothing, as no token is sent with it, meaning the entry has no change/edit. So no need to request for changed data from element-api.

3) Now we edit the entry while the live preview is still open. The ID displayed in the CP remains 84589, but underwater there now is a draft with a different ID: 88488 (we don't save this change yet, keep it a draft for now and leave the preview open)

4) Because of the edit and the live preview still open, Craft reloads the url in the live-preview iframe causing the front-end to reload with an url like below, having the new draftID as prevref, because that's the ID we need to request to have the changed data:

``` text
?l=en-US&m=prevmapmarker&prevref=88488&x-craft-live-preview=uyIutac6QC&token=kWCJ7tj_oJe4xHvjEWwN3PTzxQRDzbFb
```

5) The frontend now loads the data as normal, but knows that there is an unsaved change in the data of this particular item. It knows that because there is a token applied to the url, there is a x-craft-live-preview inside the url and the prevref ID is not yet loaded. So after loading all items as usual from element-api via another action, it now also loads this particular item seperately via the special mapmarker preview action. See code of this action in my first post above.

6) This preview action in the element-api also gets the token and all, so also knows this should return the PREVIEW data, not the live data. And it worked fine before. But no anymore. It still retreives the preview draft ID from the frontend (which was sent via the preview targets `{id}` variable), but instead of returning the preview data for this entry, it now returns:

![image](https://user-images.githubusercontent.com/69322733/131493088-61a604dd-225d-4dc1-ada1-d54d8031364e.png)

Causing the system to not function anymore, while this was working fine before.

So again, please fix this in Craft. There's nothing mentioned about this in the upgrade guide I was working through closely. There's nothing changed in our code and we didn't upgrade element-api plugin either. Only Craft was upgraded several times.

**Comment by brandonkelly** [comment]
I closed the issue because it’s still working in the way we recommend, so it’s not a Craft bug. That doesn’t mean I’m not going to help you though.

Thanks for the detailed explanation of what’s going on. It wasn’t clear to me that the draft data is meant to be loaded _in addition to_ the normal entry, but that answers that.

One of the main changes in 3.7 was the addition of _provisional drafts_ – a special classification for drafts which were automatically created while you are editing an entry. You can see #7899 for a full explanation, and #6681 for the discussion that inspired the change.

Because provisional drafts are meant to be hidden from everyone except the user that they were created for, they do not show up in entry draft queries by default. You must explicitly call `provisionalDrafts()` on the query to include them in the results.

If you’re not sure whether you’re looking for a provisional draft or not, you can pass `null` to the function, which means both normal and provisional drafts will be included:

```php
$id = Craft::$app->request->getQueryParam('prevref');

$draft = Entry::find()
    ->id($id)
    ->drafts()
    ->provisionalDrafts(null)
    ->anyStatus()
    ->one();
```

Add that one new param to your draft query within your mapmarker code, and you should be good to go.

**Comment by bencresty** [comment]
> I closed the issue because it’s still working in the way we recommend, so it’s not a Craft bug. That doesn’t mean I’m not going to help you though.

Great that you'd like to help. There is no document about a 'recommended way of doing things' for Craft AFAIK though. Documentation like that for Craft is hard to find on the website to begin with, most things I know now had to be figured out basically by just trial and error or, when not stuck, answering on the Discord channel, because the official documentation, especially on the php side of things, is very tiny and is lacking lots of important stuff. Also searching on google most of the time isn't returning much useful answers, or stuff that's outdated because things tend to change a lot.

And again, the code that was working for months and wasn't particularly amaturistic, but done well and even seemed the only way to accomplish at the time.

> Thanks for the detailed explanation of what’s going on. It wasn’t clear to me that the draft data is meant to be loaded _in addition to_ the normal entry, but that answers that.

I'd say it makes a lot of sense to be able to request data including the preview data when in live preview mode and working headlessly. How else could we show a customer a preview that's actually showing current data the customer wants to preview?

> One of the main changes in 3.7 was the addition of _provisional drafts_ – a special classification for drafts which were automatically created while you are editing an entry. You can see #7899 for a full explanation, and #6681 for the discussion that inspired the change.

It would have been nice if this was documented in the upgrade guide. Things like these really should be highlighted in there as this breaks applications and it's obviously not easy to find out what happened when stuff like this breaks because of such a change. Which is proven by the fact that we are chatting about this for days now and even you only now bring this up.

So please be more considerate and with next upgrades and the upgrade guide. This has taken me many hours now, also to explain things here, and a lot of headages. Especially because an upcoming client project here will need this stuff too, which was already build and ready to go because done in this previous project.

While just upgrading Craft on an existing project that worked flawlessly. And even doing it carefully and reading everything I could from the upgrade guide. This is no joke and costed a lot of unneeded time and money.

> Because provisional drafts are meant to be hidden from everyone except the user that they were created for, they do not show up in entry draft queries by default. You must explicitly call `provisionalDrafts()` on the query to include them in the results.

But wasn't that the whole point of sending the token to the element-api before? Sending the token to the element-api previously made us get the preview data. 

> Add that one new param to your draft query within your mapmarker code, and you should be good to go.

That's great, but as written before we're using element-api.php (see element-api.php 'action'-code in the opening post), so it's not clear to me how to enter this new `provisionalDrafts` function in there without being hacky.
I just tried `'provisionalDrafts' => false` and `'provisionalDrafts' => null` but that doesn't do a thing.

At least great that we finally know what caused the issue now. We were starting to worry.

Only question left now: How can we enter this `provisionalDrafts` addition within the 'criteria' of the element-api the right way, so without being too hacky or doing double querying?

Thanks

**Comment by brandonkelly** [comment]
> There is no document about a 'recommended way of doing things' for Craft AFAIK though.

It’s documented here: https://craftcms.com/docs/3.x/entries.html#previewing-decoupled-front-ends

> How else could we show a customer a preview that's actually showing current data the customer wants to preview?

If you follow the suggestion in the documentation or what I wrote in my first reply, it will work without any special back-end code. The requested entry will be automatically swapped out with the entry being previewed (whether that’s a normal draft or a provisional draft – or a past revision.)



> But wasn't that the whole point of sending the token to the element-api before? Sending the token to the element-api previously made us get the preview data.

The token tells Craft that it should swap out the requested (canonical) entry with the draft/revision, whenever it is queried. As far as I can tell, you’re not actually taking advantage of that functionality as you’re querying the draft directly by its own ID. It’s hard to say for sure without seeing the rest of your mapmarker code, but my hunch is that passing the token isn’t currently doing anything for you in this case.

> That's great, but as written before we're using element-api.php (see element-api.php 'action'-code in the opening post), so it's not clear to me how to enter this new `provisionalDrafts` function in there without being hacky.
> I just tried `'provisionalDrafts' => false` and `'provisionalDrafts' => null` but that doesn't do a thing.
> …
> Only question left now: How can we enter this `provisionalDrafts` addition within the 'criteria' of the element-api the right way, so without being too hacky or doing double querying?

You would set `provisionalDrafts => null` wherever you are currently fetching the draft, via the `prevref` query param (supposedly in your mapmarker code?). I don’t think you are fetching the draft from the main endpoint `criteria`; otherwise there would already be a `drafts`/`draftId`/`draftOf` param in there.

Alternatively, if you follow our guidance you won’t have to worry about setting the `provisionalDrafts` param in the first place.

**Comment by bencresty** [comment]
Hi @brandonkelly , I finally got more time to fix this issue in this project while working on other projects in the meantime. I solved it by getting rid of the special workflow that was needed in the previous Craft versions and now let Craft do the relation checking between the draftID and the live ID of that entry. Besides this now I only needed to use the `canonicalId` in the preview target url, instead of the `id` to make sure the preview got loaded. So with this it's working again with the latest Craft versions!

It wasn't particularly fun that this needed to be changed and it costed a lot of time, but at least it's working now. Thanks for your quick responses!
